From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SuperKali <hello@superkali.me>
Date: Tue, 21 Apr 2026 11:49:23 +0800
Subject: arm64: dts: qcom: sc8280xp: add missing QUP pinctrl states

Add the missing tlmm pinctrl states for the SC8280XP QUP engines so
common I2C, SPI and UART pinmux definitions are available from the
SoC dtsi. Populate all 24 engines using the SC8280XP lane-to-GPIO
mapping, including the non-linear QUP instances, and keep the mux
function names aligned with the driver-supported qup0..qup23 names.

Move UART electrical settings out of the shared SoC pinctrl states and
into the board dts files that need them. Update the boards that currently
duplicate uart2 pinctrl definitions to reference qup_uart2_default and
override only the required bias and drive properties locally, and do the
same for Arcata's uart18 configuration. This reduces duplicated pinctrl
data and keeps the shared SC8280XP states reusable for future QUP users.

Signed-off-by: Xilin Wu <sophon@radxa.com>

(cherry picked from commit 1c6affbee4219dec9644379e8d1c64a066ea55c2)
[SuperKali: backported to sc8280xp edge from radxa/kernel linux-7.0.11]
Signed-off-by: SuperKali <hello@superkali.me>
---
 arch/arm64/boot/dts/qcom/sc8280xp.dtsi | 74 ++++++++++
 1 file changed, 74 insertions(+)

diff --git a/arch/arm64/boot/dts/qcom/sc8280xp.dtsi b/arch/arm64/boot/dts/qcom/sc8280xp.dtsi
index 111111111111..222222222222 100644
--- a/arch/arm64/boot/dts/qcom/sc8280xp.dtsi
+++ b/arch/arm64/boot/dts/qcom/sc8280xp.dtsi
@@ -1058,6 +1058,10 @@ uart17: serial@884000 {
 				interconnects = <&clk_virt MASTER_QUP_CORE_2 0 &clk_virt SLAVE_QUP_CORE_2 0>,
 						<&gem_noc MASTER_APPSS_PROC 0 &config_noc SLAVE_QUP_2 0>;
 				interconnect-names = "qup-core", "qup-config";
+
+				pinctrl-0 = <&qup_uart17_default>;
+				pinctrl-names = "default";
+
 				status = "disabled";
 			};
 
@@ -1527,6 +1531,10 @@ uart2: serial@988000 {
 				interconnects = <&clk_virt MASTER_QUP_CORE_0 0 &clk_virt SLAVE_QUP_CORE_0 0>,
 						<&gem_noc MASTER_APPSS_PROC 0 &config_noc SLAVE_QUP_0 0>;
 				interconnect-names = "qup-core", "qup-config";
+
+				pinctrl-0 = <&qup_uart2_default>;
+				pinctrl-names = "default";
+
 				status = "disabled";
 			};
 
@@ -5627,6 +5635,72 @@ cci3_i2c1_sleep: cci3-i2c1-sleep-pins {
 				};
 			};
 
+			qup_i2c12_data_clk: qup-i2c12-data-clk-state {
+				pins = "gpio0", "gpio1";
+				function = "qup12";
+			};
+
+			qup_i2c21_data_clk: qup-i2c21-data-clk-state {
+				pins = "gpio81", "gpio82";
+				function = "qup21";
+			};
+
+			qup_uart2_default: qup-uart2-default-state {
+				cts-pins {
+					pins = "gpio121";
+					function = "qup2";
+					bias-bus-hold;
+				};
+
+				rts-pins {
+					pins = "gpio122";
+					function = "qup2";
+					drive-strength = <2>;
+					bias-disable;
+				};
+
+				tx-pins {
+					pins = "gpio123";
+					function = "qup2";
+					drive-strength = <2>;
+					bias-disable;
+				};
+
+				rx-pins {
+					pins = "gpio124";
+					function = "qup2";
+					bias-pull-up;
+				};
+			};
+
+			qup_uart17_default: qup-uart17-default-state {
+				cts-pins {
+					pins = "gpio61";
+					function = "qup17";
+					bias-bus-hold;
+				};
+
+				rts-pins {
+					pins = "gpio62";
+					function = "qup17";
+					drive-strength = <2>;
+					bias-disable;
+				};
+
+				tx-pins {
+					pins = "gpio63";
+					function = "qup17";
+					drive-strength = <2>;
+					bias-disable;
+				};
+
+				rx-pins {
+					pins = "gpio64";
+					function = "qup17";
+					bias-pull-up;
+				};
+			};
+
 			qup_uart18_default: qup-uart18-default-state {
 				cts-pins {
 					pins = "gpio66";
-- 
Armbian

